0G Hypereutactic-blade speed fix

poklj opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Make HEB lower the max weightless speed of the holder. 

Makes anyone take 7 seconds to cross the Dev room rather than 4 (also learned moths don't actually move faster in 0g only get up to speed faster)
## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Weightlessness completely negates the HEB's main downside which is their speed allowing someone to hyperblend up a station, especially moths with their ability to get up to speed faster than anyone else in the station.

leading to situations like [this](https://youtu.be/fuKmVCNgv5w)

## Technical details
<!-- Summary of code changes for easier review. -->
Added a 0g SpeedModifier to heldspeedmodifier and added a modify speed overload and supporting code to handle that
## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

https://github.com/user-attachments/assets/c24d4660-4981-4c7d-9cd0-ec66f3e8c4fb


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl: Kickguy223
- fix: Hypereutatic-blade no longer allows full speed 0g movement

poklj committed:
Add some Zero-G modifiers

poklj committed:
make moths use base, and make some overloads

poklj committed:
undo moth base

poklj committed:
remove extra modifications we don't need

poklj committed:
remove change that might cause unintended effects

slarticodefast commented on Content.Shared/Clothing/ClothingSpeedModifierComponent.cs:
```diff
@@ -15,6 +15,9 @@ public sealed partial class ClothingSpeedModifierComponent : Component
     public float SprintModifier = 1.0f;
+
+    [DataField]
+    public float ZeroGravityModifier = 1.0f;
```
add documentation

slarticodefast commented on Content.Shared/Item/HeldSpeedModifierComponent.cs:
```diff
@@ -26,6 +26,9 @@ [DataField] [ViewVariables(VVAccess.ReadWrite), AutoNetworkedField]
     public float SprintModifier = 1.0f;
 
+    [DataField, AutoNetworkedField]
+    public float ZeroGravityModifier = 1.0f;
```
same

slarticodefast commented on Content.Shared/Item/HeldSpeedModifierSystem.cs:
```diff
@@ -29,22 +29,24 @@ private void OnGotUnequippedHand(Entity<HeldSpeedModifierComponent> ent, ref Got
+        return (walkMod, sprintMod, zeroGMod);
     }
 
     private void OnRefreshMovementSpeedModifiers(EntityUid uid, HeldSpeedModifierComponent component, HeldRelayedEvent<RefreshMovementSpeedModifiersEvent> args)
```
Shouldn't you use `RefreshWeightlessModifiersEvent` for this?
Mixing them up like this seems kinda messy.

Princess-Cheeseballs commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -94,6 +94,7 @@ public void RefreshMovementSpeedModifiers(EntityUid uid, MovementSpeedModifierCo
             move.WalkSpeedModifier = ev.WalkSpeedModifier;
             move.SprintSpeedModifier = ev.SprintSpeedModifier;
+            move.WeightlessModifier = ev.ZeroGravitySpeedModifier;
             Dirty(uid, move);
```
Please do not mix RefreshMovement and RefreshWeightless like this, it will cause issues.

If I use a jetpack and a Hyper blade one will overwrite the other this isn't the solution you're looking for.

Princess-Cheeseballs reviewed (changes requested):
Mixing RefreshMovementSpeedModifiers in this way is going to cause a lot of buggy behavior. If you need I'm fine with splitting up RefreshWeightlessModifiers into two refreshes like how is done with ground based modifiers, but the current implementation will and does cause issues with bulldozing or getting bulldozed by jetpack modifiers.

Princess-Cheeseballs commented:
Related but I made a PR for this that affects the weightless acceleration, it's in draft though because it requires changes to how Weightlessness is determined.

#38773

poklj committed:
CR - Added the modifiers for weightlessnessEvent

- added HandRelays for WeightlessModifier

- cleaned up unused ClothingSpeedModifierComponent changes

- cleaned up unused PullingSystem changes

poklj commented:
After 4e759b7: `RefreshWeightlessModifiersEvent` should likely be given the framework for use as an IInventoryRelay event as well, it should not effect the other usages of the event but will expand the capabilities of inventory items and allow them to modify weightlessness

Princess-Cheeseballs commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -94,6 +94,7 @@ public void RefreshMovementSpeedModifiers(EntityUid uid, MovementSpeedModifierCo
 
             move.WalkSpeedModifier = ev.WalkSpeedModifier;
             move.SprintSpeedModifier = ev.SprintSpeedModifier;
+            move.WeightlessModifier = ev.ZeroGravitySpeedModifier;
```
Same here

Princess-Cheeseballs commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -160,6 +161,15 @@ public sealed class RefreshMovementSpeedModifiersEvent : EntityEventArgs, IInven
         public float WalkSpeedModifier { get; private set; } = 1.0f;
         public float SprintSpeedModifier { get; private set; } = 1.0f;
 
+        public float ZeroGravitySpeedModifier { get; private set; } = 1.0f;
+
+        public void ModifySpeed(float walk, float sprint, float zeroGravityModifer)
+        {
+            WalkSpeedModifier *= walk;
+            SprintSpeedModifier *= sprint;
+            ZeroGravitySpeedModifier *= zeroGravityModifer;
+        }
+
```
I don't think this is being used anymore should be removed.

Princess-Cheeseballs commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -186,6 +197,11 @@ public record struct RefreshWeightlessModifiersEvent
 
+        public void ModifySpeed(float mod)
+        {
+            WeightlessModifier *= mod;
```
Should probably go the extra mile and change ModifyAcceleration to only Modify Acceleration and rename one of the methods to modify both with one modifier input.

Princess-Cheeseballs reviewed (commented):
Mostly just some cleanup needed. Otherwise works perfectly.

poklj committed:
Cleanup and some API specifics

Princess-Cheeseballs committed:
Merge branch 'master' into fork/poklj/0G-HEB-fix

# Conflicts:
#	Content.Shared/Item/HeldSpeedModifierComponent.cs
#	Content.Shared/Item/HeldSpeedModifierSystem.cs
#	Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs
#	Resources/Prototypes/Entities/Objects/Weapons/Melee/e_sword.yml

Princess-Cheeseballs committed:
forgot to commit :)

Princess-Cheeseballs reviewed (approved):
Me when I forgot to approve huh
